New lint: `struct_fields_rest_default`

WeiTheShinobi opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust-clippy/pull/14412)*

Closes rust-lang/rust-clippy#14382 

changelog: [`struct_fields_rest_default`]: add new lint

WeiTheShinobi committed:
New lint: `struct_fields_rest_default`

samueltardieu commented on clippy_lints/src/struct_fields_rest_default.rs:
```diff
@@ -0,0 +1,77 @@
+
+impl EarlyLintPass for StructFieldsDefault {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
+        if let ExprKind::Struct(struct_expr) = &expr.kind
+            && let StructRest::Base(base) = &struct_expr.rest
+            && !base.span.in_external_macro(cx.sess().source_map())
+            && let ExprKind::Call(call, _) = &base.kind
+            && let ExprKind::Path(_, path) = &call.kind
+            && let [part1, part2] = path.segments.as_slice()
+            && part1.ident.name == rustc_span::sym::Default
+            && part2.ident.name == rustc_span::kw::Default
```
This will match even if the `Default` that is in scope is not the one from the standard library. If you had used a late lint, you would be able to match against the real `Default::default()` diagnostic item as paths would be resolved. Also, you would be able to match against `Foo::default()` in your tests, which is the same thing as `Default::default()`.

With a late lint, you could use something like the shorter:

```suggestion
        if let ExprKind::Struct(_, _, StructTailExpr::Base(base)) = &expr.kind
            && !base.span.in_external_macro(cx.sess().source_map())
            && let ExprKind::Call(func, _) = base.kind
            && let Some(did) = path_def_id(cx, func)
            && cx.tcx.is_diagnostic_item(sym::default_fn, did)
```

samueltardieu commented on tests/ui/auxiliary/external_macro.rs:
```diff
@@ -0,0 +1,15 @@
+#[macro_export]
+macro_rules! external_struct_rest_default {
+    () => {
+        #[derive(Default)]
+        struct ExternalDefault {
+            a: i32,
+            b: i32,
+        }
+
+        let _ = ExternalDefault {
+            a: 10,
+            ..Default::default()
+        };
+    };
+}
```
Can't you use `proc_macros::external!{ … }` in your tests, without defining a dedicated external macro?

samueltardieu commented on clippy_lints/src/struct_fields_rest_default.rs:
```diff
@@ -0,0 +1,77 @@
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
+        if let ExprKind::Struct(struct_expr) = &expr.kind
+            && let StructRest::Base(base) = &struct_expr.rest
+            && !base.span.in_external_macro(cx.sess().source_map())
```
Any reason to exclude only external macros for the base? Shouldn't also the check be excluded if `expr` comes from expansion?

samueltardieu reviewed (commented):
The current code would have a false positive for the extra test case to add in your `main` function:

```rust
    {
        struct Default;
        impl Default {
            fn default() -> Foo {
                Foo { a: 0, b: 0, c: 0 }
            }
        }
        // should not lint
        let _ = Foo { a: 0, ..Default::default() };
    }
```

Hence the proposal to switch to a late lint.

But also, why restrict this check to `default()`? Wouldn't any base value have the same issue if the goal is to force the developer to initialize all the fields?

Manishearth commented:
I'll let @samueltardieu review and do the final approval

WeiTheShinobi committed:
`struct_fields_rest_default` will reject any base value used.

WeiTheShinobi commented:
> But also, why restrict this check to `default()`? Wouldn't any base value have the same issue if the goal is to force the developer to initialize all the fields?

That makes a lot of sense! So I modified the lint to reject any base value used, and lint no need to switch to a late lint.
I also fixed the problem related to the macro check.

Thank you!

WeiTheShinobi commented:
Or we can restrict only Default::default and Foo::default,
providing flexibility for developers to use special cases.
When developers use these cases, they may have a better understanding of what they are doing,
rather than just trying to be lazy.
not lint something like
`..Foo::special_case()`
Is it make sense?

samueltardieu commented:
Indeed, right now it looks like the lint would catch many legitimate cases: <https://github.com/rust-lang/rust-clippy/actions/runs/13925788373?pr=14412>. Maybe try restricting this to `Default::default()` or `Type::default()` and see how it behaves.

Out of curiosity, do you know of any project which forbids this, or would want to? Or is it a "just in case" lint?

WeiTheShinobi commented:
> Indeed, right now it looks like the lint would catch many legitimate cases: https://github.com/rust-lang/rust-clippy/actions/runs/13925788373?pr=14412. Maybe try restricting this to `Default::default()` or `Type::default()` and see how it behaves.

Ok.

> Out of curiosity, do you know of any project which forbids this, or would want to? Or is it a "just in case" lint?

I'm not sure, but I've encountered bugs caused by this kind of situation before. I'll use it in our team's project.

Thank you!

WeiTheShinobi committed:
Restrict only `*::default`

WeiTheShinobi committed:
Fix `struct_fields_rest_default` helping message

flip1995 commented on clippy_lints/src/struct_fields_rest_default.rs:
```diff
@@ -0,0 +1,90 @@
+                    snippet(cx, base.span, "")
+                ),
+                Some(expr.span),
+                "explicitly specify all fields or use other base value instead of `..*::default()`",
```
I think this lint should only trigger, if the `Default` implementation is derived, but not when it is manually implemented. Manually implemented it isn't different from using something like `Foo::get_foo()`.

WeiTheShinobi replied:
Sure.

flip1995 commented on clippy_lints/src/struct_fields_rest_default.rs:
```diff
@@ -0,0 +1,90 @@
+                STRUCT_FIELDS_REST_DEFAULT,
+                base.span,
+                format!(
+                    "should not use `..{}` to omit rest of struct field initialization",
```
```suggestion
                    "usage of `..{}` to initialize struct fields",
```

The error message should state what the code does, not what should (not) be done. This belongs into the help/note/sugg message, as you done below.

flip1995 reviewed (commented):
Some additional comments.

WeiTheShinobi committed:
lint `struct_fields_rest_default` only detect derive `Default` now

Jarcho commented:
r? samueltardieu 

Looks like you were the one actually doing the review.
